### PR TITLE
Fix-1266: Resolve stop job command issue

### PR DESCRIPTION
**Pull request by nickImbirev**
## Proposed changes

Issue description and steps to reproduce: https://github.com/zowe/vscode-extension-for-zowe/issues/1266
Issue cause: **incorrect refresh session behaviour**.

Detailed explanation: 
AS IS:
Session is updated manually during refresh, via `Object[prop]` solution. As a result, session  was not built correctly after refresh (some default values were not used).

TO BE (implemented in the PR):
Session is updated via provided API: `ZoweExplorerApiRegister.get###Api(profileValue).getSession()`. Now, session object is correct and contains default values and applied strategies.

Additional improvements in the PR:
Util refresh function refactoring and jsdoc update accordingly.

## Types of changes

What types of changes does your code introduce to Zowe Explorer?

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [x] I have added necessary documentation (if appropriate)
- [x] Any PR dependencies have been merged and published (if appropriate)

### CODE REVIEW

File: packages/zowe-explorer/__tests__/__unit__/shared/utils.unit.test.ts [hunk 1]
```diff
@@ -104,28 +104,6 @@ describe("Shared Utils Unit Tests - Function node.labelRefresh()", () => {
     });
 });
 
-describe("Shared Utils Unit Tests - Function refreshTree()", () => {
```
Comment (nickImbirev) [inline packages/zowe-explorer/__tests__/__unit__/shared/utils.unit.test.ts:107]: I will restore these tests and enhance them slightly to cover updated refresh function

File: packages/zowe-explorer/src/utils/ProfilesUtils.ts [hunk 2]
```diff
@@ -106,25 +107,29 @@ export function isTheia(): boolean {
...
-export async function refreshTree(sessNode: IZoweTreeNode) {
-    const profileType = sessNode.getProfile().type;
-    const allProf = Profiles.getInstance().getProfiles(profileType);
-    const baseProf = await Profiles.getInstance().getBaseProfile();
-    for (const profNode of allProf) {
-        if (sessNode.getProfileName() === profNode.name) {
-            setProfile(sessNode, profNode.profile);
-            const combinedSessionProfile = (await Profiles.getInstance().getCombinedProfile(profNode, baseProf))
-                .profile;
-            setSession(sessNode, combinedSessionProfile);
```
Comment (nickImbirev) [inline packages/zowe-explorer/src/utils/ProfilesUtils.ts:123]: The problem was here: `setSession(sessNode, combinedSessionProfile)`, where we manually updated session values, ignoring provided API to build session properly.

File: packages/zowe-explorer/src/utils/ProfilesUtils.ts [hunk 3]
```diff
@@ -106,25 +107,29 @@ export function isTheia(): boolean {
...
+): Promise<void> => {
+    const profileType = sessionNode.getProfile().type;
+    const profileName = sessionNode.getProfileName();
+
+    const profile = profiles.loadNamedProfile(profileName, profileType);
+    sessionNode.setProfileToChoice(profile);
+
+    const baseProfile = profiles.getBaseProfile();
+    const combinedProfile = await profiles.getCombinedProfile(profile, baseProfile);
+    const session = getSessionForProfile(combinedProfile);
```
Comment (katelynienaber) [inline packages/zowe-explorer/src/utils/ProfilesUtils.ts:128]: Why are we passing in this function as a parameter? We can call it here with `Profiles.getInstance().getSessionForProfile()` and it simplifies `syncSession` params.

File: packages/zowe-explorer/src/utils/ProfilesUtils.ts [hunk 4]
```diff
@@ -106,25 +107,29 @@ export function isTheia(): boolean {
...
-    sessNode.collapsibleState = vscode.TreeItemCollapsibleState.Collapsed;
-}
+/**
+ * Function to update session and profile information in provided node
+ * @param profiles is data source to find profiles
+ * @param getSessionForProfile is a function to build a valid specific session based on provided profile
+ * @param sessionNode is a tree node, containing session information
+ */
+type SessionForProfile = (profile: IProfileLoaded) => Session;
+export const syncSession = (profiles: Profiles) => (getSessionForProfile: SessionForProfile) => async (
```
Comment (katelynienaber) [inline packages/zowe-explorer/src/utils/ProfilesUtils.ts:117]: Can we call it `syncSessionNode` just to be clear? When I see `syncSession` I'm thinking to myself, sync with what?

I know we're passing in a sessionNode, but we could be making some other thing match the info on the session node, and not vice versa.

File: packages/zowe-explorer/src/utils/ProfilesUtils.ts [hunk 5]
```diff
@@ -106,25 +107,29 @@ export function isTheia(): boolean {
...
-    }
-    sessNode.collapsibleState = vscode.TreeItemCollapsibleState.Collapsed;
-}
+/**
+ * Function to update session and profile information in provided node
+ * @param profiles is data source to find profiles
+ * @param getSessionForProfile is a function to build a valid specific session based on provided profile
+ * @param sessionNode is a tree node, containing session information
+ */
+type SessionForProfile = (profile: IProfileLoaded) => Session;
```
Comment (katelynienaber) [inline packages/zowe-explorer/src/utils/ProfilesUtils.ts:116]: Hehe there is some complication in ZE with IProfile, IProfileLoaded, Session...Imperative has a bunch of different cred data types that are slightly different and have slightly different keys. It's a pain...maybe we can fix it sometime.

Not something to fix on this PR, just drawing your attention to it.

File: packages/zowe-explorer/src/uss/ZoweUSSNode.ts [hunk 6]
```diff
@@ -171,7 +171,9 @@ export class ZoweUSSNode extends ZoweTreeNode implements IZoweUSSTreeNode {
                 this.label,
                 localize("getChildren.error.response", "Retrieving response from ") + `uss-file-list`
             );
-            await refreshTree(sessNode);
+            await syncSession(Profiles.getInstance())((profileValue) =>
+                ZoweExplorerApiRegister.getUssApi(profileValue).getSession()
```
Comment (katelynienaber) [inline packages/zowe-explorer/src/uss/ZoweUSSNode.ts:175]: I'm not super into passing in the getSession function. I guess it makes it more extensible, but then we have to import ZoweExplorerApiRegister everywhere that we want to refresh the session node. I recall importing this class in Profiles (I think it was?) created a circular dependency at one point that was really nasty to fix.

If your tests are all passing I guess it's fine. I'm worried that I am recalling the cause of that error inaccurately. It was def related to importing some API-related class into Profiles. But it's just a thought.

**Review by katelynienaber — CHANGES_REQUESTED** [review]
Thank you for fixing this! It works fine for me, just a few comments.

File: packages/zowe-explorer/src/utils/ProfilesUtils.ts (hunk 3 above)
Comment (nickImbirev) [reply packages/zowe-explorer/src/utils/ProfilesUtils.ts:128]: So, we have two reasons for it:

1. Dependency injection (once you pass your dependencies in the constructor or in function arg(s)). After this - you can test the function easily without any complicated mocks, and also, you have control over the function behaviour.
2. This `refresh` function is super abstract, so we don't know, what type of API do we need to call (Uss, Mvs or anything else). So, the context has to be passed inside the function.

I hope it makes sense...

**Comment by nickImbirev** [comment]
Closed in favor of ZE branch

**Pull request closed without merging**